### PR TITLE
SNOW-608095: Create new SQLAlchemy Jenkins test pipeline

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-608095

The current implementation lacks the ability to specify connector install repo/wheel url, it will install the connector from pypi.
As Mark has pointed out earlier that this could be a CI design problem, I'm still looking into how to integrate with connector build job. 

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: ci/docker/sqlalchemy_build/Dockerfile [hunk 1]
```diff
@@ -5,15 +5,14 @@ FROM $BASE_IMAGE
 # This is to solve permission issue, read https://denibertovic.com/posts/handling-permissions-with-docker-volumes/
 ARG GOSU_URL=https://github.com/tianon/gosu/releases/download/1.14/gosu-amd64
 ENV GOSU_PATH $GOSU_URL
-RUN curl -o /usr/local/bin/gosu -SL $GOSU_PATH \
-    && chmod +x /usr/local/bin/gosu
+RUN curl -o /usr/local/bin/gosu -SL $GOSU_PATH
+RUN chmod +x /usr/local/bin/gosu
```
Comment (sfc-gh-mkeller) [inline ci/docker/sqlalchemy_build/Dockerfile:9]: Please revert this, less Docker layers are preferred

File: ci/log_analyze_setup.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+export CLIENT_KNOWN_SSM_FILE_PATH=$CLIENT_LOG_DIR_PATH/rt_jenkins_log_known_ssm.txt
+echo "[INFO] CLIENT_KNOWN_SSM_FILE_PATH=$CLIENT_KNOWN_SSM_FILE_PATH"
+echo "[INFO] CLIENT_KNOWN_SSM_FILE_PATH_DOCKER=$CLIENT_KNOWN_SSM_FILE_PATH_DOCKER"
+
+# [required envs]
+# To close log analyze, just set ENABLE_CLIENT_LOG_ANALYZE to not "true", e.g. "false".
+export ENABLE_CLIENT_LOG_ANALYZE="true"
+
+# The environment variable used by log analyze module
+export CLIENT_DRIVER_NAME=PYTHON
```
Comment (sfc-gh-mkeller) [inline ci/log_analyze_setup.sh]:
```suggestion
export CLIENT_DRIVER_NAME=SQLALCHEMY
```

File: ci/test_docker.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+arch=$(uname -p)
+
+echo "[Info] Building docker image"
+if [[ "$arch" == "aarch64" ]]; then
+  BASE_IMAGE=$BASE_IMAGE_MANYLINUX2014AARCH64
+  GOSU_URL=https://github.com/tianon/gosu/releases/download/1.14/gosu-arm64
+else
+  BASE_IMAGE=$BASE_IMAGE_MANYLINUX2014
+  GOSU_URL=https://github.com/tianon/gosu/releases/download/1.14/gosu-amd64
+fi
```
Comment (sfc-gh-mkeller) [inline ci/test_docker.sh:32]:
```suggestion
GOSU_URL=https://github.com/tianon/gosu/releases/download/1.14/gosu-arm64
if [[ "$arch" == "aarch64" ]]; then
  BASE_IMAGE=$BASE_IMAGE_MANYLINUX2014AARCH64
else
  BASE_IMAGE=$BASE_IMAGE_MANYLINUX2014
fi
```

File: ci/test_docker.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+echo "[Info] Start building docker image"
+docker build --pull -t ${CONTAINER_NAME}:1.0 --build-arg BASE_IMAGE=$BASE_IMAGE  --build-arg GOSU_URL="$GOSU_URL" . -f Dockerfile
```
Comment (sfc-gh-mkeller) [inline ci/test_docker.sh:35]: Very recently I found a better way to build and consume Docker images temporarily. Please remove this line and modify the docker run command similar to https://github.com/snowflakedb/snowflake-sqlalchemy/blob/5d17bfb3dbfb1a9b29d1156c1da538ecf61847e9/ci/build_docker.sh#L24

This way the docker image is removed after the run command finishes and the built layers are not kept around.

File: ci/test_linux.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+#   - Versions to be tested should be passed in as the first argument, e.g: "3.7 3.8". If omitted 3.7-3.10 will be assumed.
+#   - This script assumes that ../dist/repaired_wheels has the wheel(s) built for all versions to be tested
+#   - This is the script that test_docker.sh runs inside of the docker container
+
+PYTHON_VERSIONS="${1:-3.7 3.8 3.9 3.10}"
+THIS_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+SQLALCHEMY_DIR="$( dirname "${THIS_DIR}")"
+
+# Install one copy of tox
+python3 -m pip install -U tox tox-external-wheels
```
Comment (sfc-gh-mkeller) [inline ci/test_linux.sh:14]: We've moved SQLAlchemy to use tox4, so `tox-external-wheels` is not necessary
```suggestion
python3 -m pip install --pre tox
```

File: tox.ini [hunk 6]
```diff
@@ -12,11 +12,32 @@ extras =
...
+    CLIENT_LOG_DIR_PATH_DOCKER
     PYTEST_ADDOPTS
-setenv = COVERAGE_FILE = {env:COVERAGE_FILE:{toxworkdir}/.coverage.{envname}}
+setenv =
+    COVERAGE_FILE = {env:COVERAGE_FILE:{toxworkdir}/.coverage.{envname}}
+    ci: SNOWFLAKE_PYTEST_OPTS = -vvv
+    # Add common parts into pytest command
+    SNOWFLAKE_PYTEST_COV_LOCATION = {env:JUNIT_REPORT_DIR:{toxworkdir}}/junit.{envname}-{env:cloud_provider:dev}.xml
+    SNOWFLAKE_PYTEST_COV_CMD = --cov snowflake.sqlalchemy --junitxml {env:SNOWFLAKE_PYTEST_COV_LOCATION} --cov-report=
+    SNOWFLAKE_PYTEST_CMD = pytest {env:SNOWFLAKE_PYTEST_OPTS:} {env:SNOWFLAKE_PYTEST_COV_CMD}
```
Comment (sfc-gh-mkeller) [inline tox.ini:39]: You added all these things, but you never use the final `SNOWFLAKE_PYTEST_CMD`

File: tox.ini [hunk 7]
```diff
@@ -12,11 +12,32 @@ extras =
     development
     pandas
 deps = pip
+external_wheels =
+    py37-ci: dist/*.whl
+    py38-ci: dist/*.whl
+    py39-ci: dist/.whl
+    py310-ci: dist/.whl
```
Comment (sfc-gh-mkeller) [inline tox.ini:19]: Since we switched to using tox4, we don't use `tox-external-wheels` anymore
```suggestion
```

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

File: ci/test_docker.sh (hunk 3 above)
Comment (sfc-gh-aling) [reply ci/test_docker.sh:32]: qq: this part is copied from connector, the gosu release seems to be platform dependent -- arm vs amd?

File: ci/docker/sqlalchemy_build/Dockerfile (hunk 1 above)
Comment (sfc-gh-aling) [reply ci/docker/sqlalchemy_build/Dockerfile]: reverted

File: ci/log_analyze_setup.sh (hunk 2 above)
Comment (sfc-gh-aling) [reply ci/log_analyze_setup.sh]: I have removed the this file first since it's not used in the sql alchemy test

File: ci/test_docker.sh (hunk 4 above)
Comment (sfc-gh-aling) [reply ci/test_docker.sh]: nice finding, updated

File: tox.ini (hunk 6 above)
Comment (sfc-gh-aling) [reply tox.ini]: removed unused env vars

File: tox.ini (hunk 7 above)
Comment (sfc-gh-aling) [reply tox.ini]: updated

File: ci/test_docker.sh [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    -e PIP_DISABLE_PIP_VERSION_CHECK=1 \
+    -e OPENSSL_FIPS=1 \
+    -e LOCAL_USER_ID=${user_id} \
+    -e AWS_ACCESS_KEY_ID \
+    -e AWS_SECRET_ACCESS_KEY \
+    -e SF_REGRESS_LOGS \
+    -e SF_PROJECT_ROOT \
+    -e cloud_provider \
+    -e JENKINS_HOME \
+    -e is_old_driver \
```
Comment (sfc-gh-mkeller) [inline ci/test_docker.sh:48]:
```suggestion
```

File: ci/test_docker.sh (hunk 3 above)
Comment (sfc-gh-mkeller) [reply ci/test_docker.sh:32]: Ahh, I missed that last part of the URLs

**Review by sfc-gh-mkeller — APPROVED** [review]

File: ci/test_docker.sh (hunk 8 above)
Comment (sfc-gh-aling) [reply ci/test_docker.sh]: fixed

**Pull request merged**